server/src/main/java/org/elasticsearch/cluster/routing/RoutingNode.java

```diff
@@ -142,11 +147,10 @@ void add(ShardRouting shard) {
             relocatingShards.add(shard);
         }
         shardsByIndex.computeIfAbsent(shard.index(), k -> new HashSet<>()).add(shard);
-        assert invariant();
+        assert validate == false || invariant();
     }
 
     void update(ShardRouting oldShard, ShardRouting newShard) {
-        assert invariant();
```

Comment:
The invariant asserted **after** operation as well. I do not think it is required to do it twice.

Reply:
Nice!